persist/sqlite/contracts.go

```diff
@@ -864,11 +864,52 @@ raw_revision, host_sig, renter_sig, confirmed_revision_number, contract_status,
 	return
 }
 
+// v2ContractExists is a helper that checks if a contract already exists.
+//
+// If the contract exists, and its contract_status is rejected the
+// existing contract will be deleted so the new contract can be inserted
+// and returns (nil).
+//
+// If the contract exists, and its contract_status is not rejected,
+// [contracts.ErrContractExists] will be returned.
+//
+// If the contract does not exist, it returns (nil).
+
+func v2ContractExists(tx *txn, contractID types.FileContractID) error {
+	var dbID int64
+	var status contracts.V2ContractStatus
+	err := tx.QueryRow(`SELECT id, contract_status FROM contracts_v2 WHERE contract_id=$1;`, encode(contractID)).Scan(&dbID, &status)
+	if errors.Is(err, sql.ErrNoRows) {
+		return nil // contract does not exist
+	} else if err != nil {
+		return fmt.Errorf("failed to check contract existence: %w", err)
+	}
+
+	if status != contracts.V2ContractStatusRejected {
+		return contracts.ErrContractExists
+	}
+
+	_, err = tx.Exec(`DELETE FROM contracts_v2 WHERE id=$1;`, dbID)
```

Comment:
Did you check that this won't fail due to any of the foreign key constraints? The state elements set the reference to NULL implicitly and `renewed_to` / `renewed_from` too but that still leaves the sector roots and account funding. The roots are probably fine because the sectors get pruned but what about the account funding?